examples/greeter-extension/README.md

```diff
@@ -0,0 +1,328 @@
+  const greeters = await this.greeters();
+  // Find a greeter that can speak the given language
+  for (const greeter of greeters) {
+    if (greeter.language === language) {
+      greeting = greeter.greet(name);
+      break;
+    }
+  }
+  if (!greeting) {
+    // Fall back to English
+    greeting = `Hello, ${name}`;
+  }
+  if (this.options && this.options.color) {
+    greeting = chalk.keyword(this.options.color)(greeting);
+  }
+  return greeting;
+}
+```
+
+## Implement an extension
+
+Modules that want to connect to `greeter` extension point must implement
+`Greeter` interface in their extension. The key attribute is that the
+`GreeterExtensionPoint` being extended knows nothing about the module that is
+connecting to it beyond the scope of that contract. This allows `greeters` built
+by different individuals or companies to interact seamlessly, even without their
+knowing much about one another.
+
+```ts
+import {Greeter, asGreeter} from '../types';
+import {bind, inject} from '@loopback/context';
+
+/**
+ * Options for the Chinese greeter
+ */
+export interface ChineseGreeterOptions {
+  // Name first, default to `true`
+  nameFirst: boolean;
+}
+
+/**
+ * A greeter implementation for Chinese
+ */
+@bind(asGreeter)
+export class ChineseGreeter implements Greeter {
+  language = 'zh';
+
+  constructor(
+    /**
+     * Inject the configuration for ChineseGreeter
+     */
+    @inject('greeters.ChineseGreeter.options', {optional: true})
+    private options: ChineseGreeterOptions = {nameFirst: true},
+  ) {}
+
+  greet(name: string) {
+    if (this.options && this.options.nameFirst === false) {
+      return `你好，${name}！`;
+    }
+    return `${name}，你好！`;
+  }
+}
+```
+
+Please note we use
+[`@bind`](https://loopback.io/doc/en/lb4/Binding.html#configure-binding-attributes-for-a-class)
+to customize how the class can be bound. In this case, `asGreeter` is a binding
+template function, which is equivalent as configuring a binding with
+`{extensionPoint: 'greeter'}` tag and in the `SINGLETON` scope.
+
+```ts
+/**
+ * A binding template for greeter extensions
+ * @param binding
+ */
+export const asGreeter: BindingTemplate = binding =>
+  binding.inScope(BindingScope.SINGLETON).tag({extensionPoint: 'greeter'});
+```
+
+## Register an extension point
+
+To register an extension point, we simply bind the implementation class to a
+`Context`. For example:
+
+```ts
+app
+  .bind('greeter-extension-point')
+  .toClass(GreeterExtensionPoint)
+  .inScope(BindingScope.SINGLETON);
+```
+
+**NOTE**: Your extension point may choose to use a different binding scope.
+
+The process can be automated with a component:
+
+```ts
+import {createBindingFromClass} from '@loopback/context';
+import {Component} from '@loopback/core';
+import {GreeterExtensionPoint} from './greeter-extension-point';
+
+export class GreeterComponent implements Component {
+  bindings = [
+    createBindingFromClass(GreeterExtensionPoint, {
+      key: 'greeter-extension-point',
```

Comment:
Is there any particular reason for using manually constructed binding key in camel case, instead of using extension-point-class-name as the binding key?

```ts
createBindingFromClass(GreeterExtensionPoint, {
  key: 'GreeterExtensionPoint',
});
```

Ideally, I'd like LB4 to take care of the binding key for me.

```ts
createBindingFromClass(GreeterExtensionPoint)
```

Alternatively, we can provide `asExtensionPoint` template. I am not sure what's the right way for creating a binding from a class using a template, is it the following?

```ts
createBindingFromClass(GreeterExtensionPoint)
  .apply(asExtensionPoint);
```

Reply:
Fixed.